proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1569 @@
+- [Export Schema in OpenAPI v3 format](#use-case-exporting-schema-in-openapi-format)
+- [Validate an output document](#use-case-validating-an-output-document)
+
+The following sections detail each use case, in turn...
+
+#### Use Case: Required input
+
+When the author cannot/will not provide a default value for a data value and requires the consumer to provide it.
+
+Where possible, authors are encouraged to use a type-specific "non-empty" validation. As a fallback, they may opt for nullable+not_null.
+
+##### Required input for types with length
+
+**String Values**
+
+The recommended way for an author to indicate required string input is to expect a non-empty string:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@schema/validation min_len=1
+  namespace: ""
+  #@schema/validation min_len=1
+  username: ""
+```
+
+This case is so common (it is our understanding that most string-based data values are meant to be non-empty), we offer this convenience:
+
+```yaml
+#@data/values-schema
+#@schema/validation-defaults-for-strings min_len=1
+---
+dex:
+  namespace: ""
+  username: ""
+```
+
+Which is equivalent to the prior example, but applies the default to any descendent that contains a string-typed value. If a particular node already has an `@assert/validate` annotation, it is skipped.
+
+**Array Values**
+
+Similarly, a non-empty value of an array has at least one element:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  oauth2:
+    #@schema/validation min_len=1
+    responseTypes:
+    - ""
+```
+
+Array-typed values are less common. And when present are not necessarily _as_ common to require a non-empty value as string-typed values. So, there is no corresponding shorthand annotation for requiring a non-empty array.
+
+
+##### Required input for types with empty value
+
+For some integer-typed (or float-typed) data values, a zero value may be invalid and thus can represent "empty".
+
+```yaml
+#@data/values-schema
+dex:
+  #@schema/validation min=1024, max=65535
+  port: 0
+```
+
+Authors are encouraged to favor this approach over [marking the data values as nullable](#required-input-for-types-with-no-empty-value) to keep their schema as simple as possible.
+ 
+##### Required input for types with no empty value
+
+Some types have no natural (or easily detectable) "empty" value. In these cases, authors can force the consumer to supply a value by marking the value as "nullable" and constraint it to be "not null":
+
+```yaml
+#@data/values-schema
+---
+#@schema/nullable
+#@schema/validation not_null=("Cloud credentials are required.", True)
+credential:
+  name:
+  secretContents:
+    cloud:
+```
+
+_(also considered: [`@schema/required`](#schemarequired))._
+
+#### Use Case: Enumeration
+
+Values can be constrained to be one of a finite site (i.e. of an enumeration).
+
+```yaml
+#@data/values-schema
+---
+volumeSnapshotLocation:
+  spec:
+    #@schema/validation one_of=["aws", "azure", "vsphere"]
+    provider: ""
+```
```

Comment:
Are enumerations constrained to a specific data type (e.g. strings) or could they hold any valid JSON/YAML data type?

Reply:
Good question. I think we'd rely on the (semi-newly-minted) [`assert.equals()`](https://carvel.dev/ytt/docs/v0.40.0/lang-ref-ytt/#assert). Which means: yes — any value YAML data type (including arrays and maps) could be members of this enumeration.

Ooof! and the docs for that module could use some elaboration. The implementation serializes both arguments and compares those serializations: https://github.com/vmware-tanzu/carvel-ytt/blob/b50fd584133ce57a147c9fa050bd4acea684a259/pkg/yttlibrary/assert.go#L30-L61

Reply:
I see, that makes total sense!

Reply:
Thank you for your question. This was a detail we hadn't hashed out. It's help to have a good idea of approach going into things like this. 🙏🏻